clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java

```diff
@@ -0,0 +1,224 @@
+                new FindCoordinatorRequestHandler(),
+                null);
+    }
+
+    /**
+     * Mark the current coordinator null.
+     * @param cause why the coordinator is marked unknown.
+     * @param currentTimeMs the current time in ms.
+     */
+    protected void markCoordinatorUnknown(final String cause, final long currentTimeMs) {
+        if (this.coordinator != null) {
+            log.info("Group coordinator {} is unavailable or invalid due to cause: {}. "
+                            + "Rediscovery will be attempted.", this.coordinator, cause);
+            this.coordinator = null;
+            timeMarkedUnknownMs = currentTimeMs;
+        } else {
+            long durationOfOngoingDisconnect = Math.max(0, currentTimeMs - timeMarkedUnknownMs);
+            if (durationOfOngoingDisconnect > this.rebalanceTimeoutMs)
+                log.debug("Consumer has been disconnected from the group coordinator for {}ms",
+                        durationOfOngoingDisconnect);
+        }
+    }
+
+    private void onSuccessfulResponse(final FindCoordinatorResponseData.Coordinator coordinator) {
+        // use MAX_VALUE - node.id as the coordinator id to allow separate connections
+        // for the coordinator in the underlying network client layer
+        int coordinatorConnectionId = Integer.MAX_VALUE - coordinator.nodeId();
+
+        this.coordinator = new Node(
+                coordinatorConnectionId,
+                coordinator.host(),
+                coordinator.port());
+        log.info("Discovered group coordinator {}", coordinator);
+        coordinatorRequestState.reset();
+    }
+
+    private void onFailedCoordinatorResponse(
+            final Exception exception,
+            final long currentTimeMs) {
+        coordinatorRequestState.updateLastFailedAttempt(currentTimeMs);
+        markCoordinatorUnknown("coordinator unavailable", currentTimeMs);
+
+        if (exception == Errors.GROUP_AUTHORIZATION_FAILED.exception()) {
+            log.debug("FindCoordinator request failed due to authorization error {}", exception.getMessage());
+            errorHandler.handle(GroupAuthorizationException.forGroupId(this.groupId));
+            return;
+        }
+
+        if (exception instanceof RetriableException) {
+            log.debug("FindCoordinator request failed due to retriable exception", exception);
+            return;
+        }
+
+        log.warn("FindCoordinator request failed due to fatal exception", exception);
+        errorHandler.handle(exception);
+    }
+
+    /**
+     * Handles the response and exception upon completing the {@link FindCoordinatorRequest}. This is invoked in the callback
+     * {@link FindCoordinatorRequestHandler}. If the response was successful, a coordinator node will be updated. If the
+     * response failed due to errors, the current coordinator will be marked unknown.
+     * @param currentTimeMs current time ins ms.
+     * @param response the response, can be null.
```

Comment:
Perhaps we could clarify this? It looks like the code expects the response to be non-null if the exception is null?